### PR TITLE
[WebAuthn] Implement Set Pin for security keys

### PR DESCRIPTION
**Pull request by pascoej**
#### 2968de52d0664b1fd91876cf47cd91a02faa6fb5
<pre>
[WebAuthn] Implement Set Pin for security keys
<a href="https://bugs.webkit.org/show_bug.cgi?id=269083">https://bugs.webkit.org/show_bug.cgi?id=269083</a>
<a href="https://rdar.apple.com/122660610">rdar://122660610</a>

Reviewed by Brent Fulgham.

This change implements setting a pin for security keys. This is accomplished by asking the user
for a new pin whenever presented with an operation that requires a security key to have user
verification setup, but it does not.

The minimum pin length is fetched from the authenticator, or a default of 4 defined in the
CTAP spec. The maximum length is defined by the spec to be 63. We bubble up errors to show
in the prompt whenever an entered PIN did not meet requirements.

The way setting a pin works is as follows: Whenever an operation requires UV, but it is not setup,
first authenticator selection is performed to confirm a authenticator is desired for the operation.
For CTAP2.1 keys, the authenticatorSelection command is used, on older FIDO2.0 authenticators, a non-rk
makeCredential with zero length pinAuth is used instead. Then, whenever an authenticator without uv is
chosen, a new pin is collected from the user and setup on the security key, which is then able to perform
the operation requiring uv.

The SetPin code is covered by unit tests. I have manually tested setting up a PIN on both CTAP2.1 and CTAP2.0
security keys.

* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.cpp:
(fido::AuthenticatorGetInfoResponse::setMinPINLength):
(fido::encodeAsCBOR):
* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::encodeBogusRequestForAuthenticatorSelection):
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::convertStringToProtocolVersion):
(fido::readCTAPGetInfoResponse):
* Source/WebCore/Modules/webauthn/fido/FidoConstants.cpp:
(fido::isCtap2Protocol):
(fido::toString):
* Source/WebCore/Modules/webauthn/fido/FidoConstants.h:
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::SetPinRequest::pinAuth const):
(fido::pin::encodeAsCBOR):
* Source/WebCore/Modules/webauthn/fido/Pin.h:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanelClient.h:
(API::WebAuthenticationPanelClient::requestNewPin const):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/WebAuthentication/Authenticator.h:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::requestNewPin):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm:
(WebKit::AuthenticatorPresenterCoordinator::updatePresenter):
(WebKit::AuthenticatorPresenterCoordinator::requestNewPin):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticationPanelClient.mm:
(WebKit::WebAuthenticationPanelClient::WebAuthenticationPanelClient):
(WebKit::wkWebAuthenticationPanelUpdate):
(WebKit::WebAuthenticationPanelClient::requestNewPin const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::feedReports):
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualService.mm:
(WebKit::authenticatorInfoForConfig):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticationFlags.h:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::isUVSetup const):
(WebKit::CtapAuthenticator::continueSetupPinAfterCommand):
(WebKit::CtapAuthenticator::continueSetupPinAfterGetKeyAgreement):
(WebKit::CtapAuthenticator::setupPin):
(WebKit::CtapAuthenticator::performAuthenticatorSelectionForSetupPin):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.h:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapCcidDriver.cpp:
(WebKit::CtapCcidDriver::transact):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapDriver.h:
(WebKit::CtapDriver::isCtap2Protocol const):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapHidDriver.cpp:
(WebKit::CtapHidDriver::continueAfterChannelAllocated):
(WebKit::CtapHidDriver::cancel):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapNfcDriver.cpp:
(WebKit::CtapNfcDriver::transact):
* Source/WebKit/UIProcess/WebAuthentication/fido/FidoService.cpp:
(WebKit::FidoService::continueAfterGetInfo):
* Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp:
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetInfoResponse)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetInfoResponse2)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestReadGetInfoResponseDeviceYubikey5c)):
(TestWebKitAPI::TEST(CTAPResponseTest, TestSerializeGetInfoResponse)):

Canonical link: <a href="https://commits.webkit.org/288259@main">https://commits.webkit.org/288259@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/ed2faaa80d07d6b1115f3cd87e804a1e0c310cc8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82266 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87398 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33327 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84371 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2001 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9812 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64116 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21864 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85335 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1393 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74857 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44394 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1295 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29038 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32368 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29662 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88754 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6832 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72508 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9798 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70671 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71726 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17884 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15862 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14877 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/927 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9525 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15046 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9399 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11169 "Built successfully") | | |

### CODE REVIEW

**Comment by pascoej** [comment]
I've been told to ignore this particular style error in the past.

File: Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp [hunk 1]
```diff
@@ -220,6 +221,33 @@ Vector<uint8_t> encodeGetAssertionRequestAsCBOR(const Vector<uint8_t>& hash, con
     return cborRequest;
 }
 
+Vector<uint8_t> encodeBogusRequestForAuthenticatorSelection()
+{
+    CBORValue::MapValue cborMap;
+    cborMap[CBORValue(1)] = CBORValue(Vector<uint8_t> { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 });
```
Comment (brentfulgham) [inline Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp]: Nit: I wonder if this could just be something like:
`CBORValue(Vector<uint8_t>(0, 32));`

File: Source/WebCore/Modules/webauthn/fido/Pin.cpp [hunk 2]
```diff
@@ -301,6 +301,11 @@ const WebCore::CryptoKeyAES& SetPinRequest::sharedKey() const
     return m_sharedKey;
 }
 
+Vector<uint8_t> SetPinRequest::pinAuth() const
```
Comment (brentfulgham) [inline Source/WebCore/Modules/webauthn/fido/Pin.cpp]: Why is this being returned by copy?

File: Source/WebCore/Modules/webauthn/fido/Pin.h [hunk 3]
```diff
@@ -138,6 +137,7 @@ struct SetPinRequest {
 public:
     WEBCORE_EXPORT const WebCore::CryptoKeyAES& sharedKey() const;
     WEBCORE_EXPORT static std::optional<SetPinRequest> tryCreate(const String& newPin, const WebCore::CryptoKeyEC&);
+    WEBCORE_EXPORT Vector<uint8_t> pinAuth() const;
```
Comment (brentfulgham) [inline Source/WebCore/Modules/webauthn/fido/Pin.h]: If `pinAuth()` is backed by a Vector member, why isn't it returned by reference so we can avoid building a Vector for most uses?

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.mm [hunk 4]
```diff
@@ -30,6 +30,8 @@
 
 #import "CtapHidDriver.h"
 #import "HidConnection.h"
+#import <IOKit/hid/IOHIDDevice.h>
+#import <wtf/cf/TypeCastsCF.h>
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidService.mm]: Is this a non-unified build finding? It's fine to keep it in, but doesn't seem relevant to the patch unless you discovered a build failure on some platforms.

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 5]
```diff
@@ -456,6 +460,118 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+void CtapAuthenticator::continueSetupPinAfterCommand(Vector<uint8_t>&& data, const String& pin, Ref<WebCore::CryptoKeyEC> peerKey)
+{
+    auto error = getResponseCode(data);
+    if (error != fido::CtapDeviceResponseCode::kSuccess) {
+        CTAP_RELEASE_LOG("continueSetupPinAfterCommand: Response of setPin was not successful: %s", base64EncodeToString(data).utf8().data());
+        observer()->authenticatorStatusUpdated(WebAuthenticationStatus::PinInvalid);
+        return;
+    }
+    m_info.mutableOptions().setClientPinAvailability(AuthenticatorSupportedOptions::ClientPinAvailability::kSupportedAndPinSet);
+    auto pinUTF8 = pin::validateAndConvertToUTF8(pin);
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:478]: Is there no need to confirm `pinUTF8` has a value?

**Review by brentfulgham — APPROVED** [review]
Looks good, though please confirm you need to return the Vector by copy, and that you don't need to confirm the optional has a value.

File: Source/WebCore/Modules/webauthn/fido/FidoConstants.h [hunk 6]
```diff
@@ -170,6 +176,8 @@ constexpr size_t kHidInitNonceLength = 8;
 
 constexpr uint8_t kHidMaxLockSeconds = 10;
 
+const size_t kPINMaxSizeInBytes = 63;
```
Comment (cdumez) [inline Source/WebCore/Modules/webauthn/fido/FidoConstants.h]: `constexpr`

File: Source/WebCore/Modules/webauthn/fido/Pin.cpp [hunk 7]
```diff
@@ -301,6 +301,11 @@ const WebCore::CryptoKeyAES& SetPinRequest::sharedKey() const
     return m_sharedKey;
 }
 
+const Vector<uint8_t>& SetPinRequest::pinAuth() const
```
Comment (cdumez) [inline Source/WebCore/Modules/webauthn/fido/Pin.cpp:304]: can be inlined.

File: Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp [hunk 8]
```diff
@@ -371,6 +371,27 @@ void AuthenticatorManager::requestPin(uint64_t retries, CompletionHandler<void(c
     });
 }
 
+void AuthenticatorManager::requestNewPin(uint64_t minLength, CompletionHandler<void(const WTF::String&)>&& completionHandler)
+{
+    auto callback = [weakThis = WeakPtr { *this }, this, completionHandler = WTFMove(completionHandler)] (const WTF::String& pin) mutable {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp]: This doesn't comply with safer C++ guidelines. Please do not capture `this`, only `weakThis`

File: Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp [hunk 9]
```diff
@@ -371,6 +371,27 @@ void AuthenticatorManager::requestPin(uint64_t retries, CompletionHandler<void(c
...
+        if (!weakThis)
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp]: You should assign `weakThis` to a `protectedThis` before you use it below.

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm [hunk 10]
```diff
@@ -128,6 +128,16 @@
         m_credentialRequestHandler(nil, error.get());
         break;
     }
+    case WebAuthenticationStatus::PINTooShort: {
+        auto error = adoptNS([[NSError alloc] initWithDomain:ASCAuthorizationErrorDomain code:ASCAuthorizationErrorPINTooShort userInfo:nil]);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm]: `RetainPtr`

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm [hunk 11]
```diff
@@ -128,6 +128,16 @@
...
+        m_credentialRequestHandler(nil, error.get());
+        break;
+    }
+    case WebAuthenticationStatus::PINTooLong: {
+        auto error = adoptNS([[NSError alloc] initWithDomain:ASCAuthorizationErrorDomain code:ASCAuthorizationErrorPINTooLong userInfo:nil]);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticatorPresenterCoordinator.mm]: `RetainPtr`

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.mm [hunk 12]
```diff
@@ -24,6 +24,7 @@
  */
 
 #import "config.h"
+
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/HidConnection.mm]: Unnecessary change.

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 13]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+    auto tokenRequest = pin::TokenRequest::tryCreate(*pinUTF8, peerKey);
+    if (!tokenRequest) {
+        CTAP_RELEASE_LOG("continueSetupPinAfterCommand: Failed to create TokenRequest.");
+        receiveRespond(ExceptionData { ExceptionCode::UnknownError, "Cannot create a TokenRequest."_s });
+        return;
+    }
+
+    auto cborCmd = encodeAsCBOR(*tokenRequest);
+    CTAP_RELEASE_LOG("continueSetupPinAfterCommand: Sending %s", base64EncodeToString(cborCmd).utf8().data());
+    driver().transact(WTFMove(cborCmd), [weakThis = WeakPtr { *this }, this, tokenRequest = WTFMove(*tokenRequest)] (Vector<uint8_t>&& data) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:493]: We should not capture `this`

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 14]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+        ASSERT(RunLoop::isMain());
+        if (!weakThis)
+            return;
+        CTAP_RELEASE_LOG("continueGetPinTokenAfterRequestPin: Response %s", base64EncodeToString(data).utf8().data());
+        weakThis->continueRequestAfterGetPinToken(WTFMove(data), tokenRequest);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp]: Please use `protectedThis`

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 15]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+    if (!setPinRequest) {
+        receiveRespond(ExceptionData { ExceptionCode::UnknownError, "Cannot create a SetPinRequest."_s });
+        return;
+    }
+    m_pinAuth = setPinRequest->pinAuth();
+    auto cborCmd = encodeAsCBOR(*setPinRequest);
+    driver().transact(WTFMove(cborCmd), [weakThis = WeakPtr { *this }, pin, peerKey = WTFMove( keyAgreement->peerKey)](Vector<uint8_t>&& response) {
+        if (!weakThis)
+            return;
+        weakThis->continueSetupPinAfterCommand(WTFMove(response), pin,  WTFMove(peerKey));
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp]: `protectedThis`

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 16]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+    });
+}
+
+void CtapAuthenticator::setupPin()
+{
+    if (!observer())
+        return;
+    CTAP_RELEASE_LOG("setupPin: Requesting new pin from delegate");
+    uint64_t minLength = m_info.minPINLength().value_or(4);
+    observer()->requestNewPin(minLength, [weakThis = WeakPtr { *this }, this] (const String& pin) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:533]: No capture of `this` and use `protectedThis` inside the lambda.

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 17]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+        ASSERT(RunLoop::isMain());
+        if (!weakThis)
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp]: We should null check `protectedThis`

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 18]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+            return;
+        }
+        if (pin.sizeInBytes() > kPINMaxSizeInBytes) {
+            weakThis->observer()->authenticatorStatusUpdated(WebAuthenticationStatus::PINTooLong); // PINTooLong
+            weakThis->performAuthenticatorSelectionForSetupPin();
+            return;
+        }
+        auto cborCmd = encodeAsCBOR(pin::KeyAgreementRequest { });
+        CTAP_RELEASE_LOG("setupPin: Sending %s", base64EncodeToString(cborCmd).utf8().data());
+        driver().transact(WTFMove(cborCmd), [weakThis = WeakPtr { *this }, this, pin] (Vector<uint8_t>&& data) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp]: ditto about capturing `this`

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 19]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+void CtapAuthenticator::performAuthenticatorSelectionForSetupPin()
+{
+    CTAP_RELEASE_LOG("performAuthenticatorSelectionForSetupPin: Requesting gesture for authenticator selection");
+    if (m_info.versions().contains(ProtocolVersion::kCtap21) || m_info.versions().contains(ProtocolVersion::kCtap21Pre)) {
+        // We should perform the authenticatorSelector command
+        driver().transact(encodeEmptyAuthenticatorRequest(CtapRequestCommand::kAuthenticatorAuthenticatorSelection), [weakThis = WeakPtr { *this }, weakDriver = WeakPtr { driver() }] (Vector<uint8_t>&& response) mutable {
+            ASSERT(RunLoop::isMain());
+            if (!weakThis)
+                return;
+            weakThis->setupPin();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp]: `protectedThis` here and above:
```
if (RefPtr protectedThis = weakThis.get())
    protectedThis->setupPin();
```

File: Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp [hunk 20]
```diff
@@ -456,6 +460,124 @@ String CtapAuthenticator::aaguidForDebugging() const
...
+        });
+    } else {
+        auto zeroLengthPinAuth = encodeBogusRequestForAuthenticatorSelection();
+        // We should send a zeroLength pinAuth
+        driver().transact(WTFMove(zeroLengthPinAuth), [weakThis = WeakPtr { *this }, weakDriver = WeakPtr { driver() }] (Vector<uint8_t>&& response) mutable {
+            ASSERT(RunLoop::isMain());
+            if (!weakThis)
+                return;
+            weakThis->setupPin();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp]: `protectedThis`

File: Source/WebKit/UIProcess/WebAuthentication/fido/FidoService.cpp [hunk 21]
```diff
@@ -71,13 +73,14 @@ void FidoService::continueAfterGetInfo(WeakPtr<CtapDriver>&& weakDriver, Vector<
     if (!driver || !observer() || response.isEmpty())
         return;
 
+    CTAP_RELEASE_LOG("Got response from getInfo: %s", base64EncodeToString(response).utf8().data());
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebAuthentication/fido/FidoService.cpp:77]: Not that all these `%s` in release logging will print nothing on customer builds. May be intentional but I thought I would pointer it out.
